File: src/qat/agent/ExecProcess.java
package qat.agent;
/**
 * This class handles the execution of a single process, and manages it's output stream collection etc.
 *
 * @author webhiker
 * @version 2.3, 17 June 1999
 */
// standard Java imports
import java.lang.Thread;
import java.lang.System;
import java.lang.Math;
import java.io.*;
import java.lang.Runtime;
import java.lang.Process;
import java.lang.IllegalThreadStateException;
import java.lang.Thread;
import java.util.Properties;
import java.util.Enumeration;
import java.util.Date;

// qa tester imports
import qat.agent.ReaderThread;
import qat.common.ConsoleServer;
import qat.common.TestObject;
import qat.common.Utils;

/**
 * This class is responsible for executing a single TestObject, and starting new threads to read it's standard output
 * and error output streams.
 * It allows for killing of the started TestObject at anytime by calling the cancel() method.
 */
public class ExecProcess extends Object {
    private static final int NOT_YET_SET          = -9999;
    public  static final int TIMEDOUT_STATE       = -9998;
    private static final int ILLEGAL_THREAD_STATE = -5;
    private static final int NULL_POINTER         = -6;
    private static final int OTHER_EXCEPTION      = -7;
    private TestObject test;
    private Process process;
    private Date startTime;
    private ReaderThread stdin;
    private ReaderThread stderr;
    private int exitValue;
    private int debugLevel; // this is the level of detail we will use for debug message. 0 = none, 10 = max
    private PrintStream stdout;
    private boolean showOutput;
    private boolean interrupted;
    private String parentDetails;
    private boolean isDaemon;	
    /** 
     * This constructs a new ExecProcess with the specified TestObject
     * @param t - the TestObject that will be executed when the start() method is called.
     * @param so - the PrintWriter to use for output.
     * @param level - the detail of debug message we want.
     * @param mode - if true, the output is displayed to stdout
     * @param parentDetails - some details about the parent so we can print it
     * to the trace files.
     */
    public ExecProcess(TestObject t, PrintStream so, int level, boolean mode, String parentDetails) {
	setDaemon(false);
	setInterrupted(false);
	test = t;
	setProcess(null);
	stdout = so;
	debugLevel = level;
	showOutput = mode;
	this.parentDetails = parentDetails;
    }
    /** 
     * This method returns the TestObject associated with this Object.
     * @return the TestObject associated with this object.
     */
    public TestObject getTestObject() {
	return test;
    }
    /** 
     * This method should not be called directly, but rather via ObjectName.start() to
     * ensure the thread is started correctly.
     */
    public void start() {
	exitValue = NOT_YET_SET;
	// make sure we allow only one instance of this process to run
	if (getProcess()!=null) {
	    System.out.println("Internal error in ExecProcess - already have a process running!");
	    return;
	}
	// take a snapshot of the current time, so we can calculate the timeout value
	startTime = new Date();
	// build the environment string list from our TestObject properties
	String commandArgs[] = test.getExecuteCommand();		

	// build the environment string list from our TestObject properties
	String environment[] = new String[test.getExecuteEnvironment().size()];
	Enumeration e = test.getExecuteEnvironment().propertyNames();
	for (int i = 0; i < environment.length; i++) {
	    environment[i] = (String)e.nextElement();
	    environment[i] = environment[i]+"="+ test.getExecuteEnvironment().getProperty(environment[i]);	
	}
	PrintWriter out=null;
	try {
	    // create and write the environment file
	    out = new PrintWriter(new FileOutputStream(test.getEnvFileName()));
		    
	    debugMsg("Command arguments :"+commandArgs.length,out,1);
	    debugMsg("-------------------",out,1);
	    for (int i = 0; i < commandArgs.length; i++) {
		debugMsg(" Argument "+i+" :"+commandArgs[i],out,1);
	    }
	    debugMsg("",out,1);
		    
	    debugMsg("Agent details :",out,1);
	    debugMsg("---------------",out,1);
	    debugMsg(parentDetails,out,1);
	    debugMsg("",out,1);
		    
	    debugMsg("Process environment :",out,1);
	    debugMsg("---------------------",out,1);
	    for (int i = 0; i < environment.length; i++)
		debugMsg(environment[i],out,1);
	    debugMsg("",out,1);

	    // patch to inherit env if we don't specify one
	    if (environment.length==0)
		environment = null;

	    setProcess(Runtime.getRuntime().exec(commandArgs, environment));
		    
	    // only store traces if it's not a daemon process
	    if (!isDaemon()) {
		// now process the output and erroutput streams of this process by starting two threads
		// which monitor these streams
		stderr = new ReaderThread("StdErr",getProcess().getErrorStream(),showOutput, test.getStdErrFileName());
		debugMsg("Starting errout thread",out,5);
		stderr.start();
		stdin  = new ReaderThread("StdOutput",getProcess().getInputStream(),showOutput, test.getStdOutFileName());
		debugMsg("Starting stdout thread",out,5);
		stdin.start();
	    }
	    debugMsg("Process started at :"+(new Date()).toString(),out,1);
	}
	catch (IOException ex) {
	    ConsoleServer.debugStackTrace(ex);
	    debugMsg("IOException starting process :"+ex.getMessage(),out,5);
	    interrupt();
	}
	catch (Throwable ex) {
	    ConsoleServer.debugStackTrace(ex);
	    debugMsg("Fatal error starting process :"+ex.getMessage(),out,5);
	    interrupt();
	    // delete this temporary files if they were created, since
	    // it's an abnormal behaviour to get an error here
	    if (test.getStdOutFileName().exists())
		test.getStdOutFileName().delete();
	    if (test.getStdErrFileName().exists())
		test.getStdErrFileName().delete();
	    if (test.getEnvFileName().exists())
		test.getEnvFileName().delete();
	}
		
	finally {
	    try {
		out.flush();
		out.close();
	    }
	    catch (Exception ex) {
		ConsoleServer.debugMsg("Error closing trace file :"+ex.getMessage(),2);
	    }
	}
    }
    
    private Process getProcess() {
	return this.process;
    }
    
    private void setProcess(Process p) {
	this.process = p;
    }
	
    /**
     * This function returns the exit value of the running process if it has finished 
     * else it returns a negative value.
     * This method will not block if the process is still running.
     * @return the absolute value of the exit value of the process, or a negative int to indicate other
     * thread states (ILLEGAL_THREAD_STATE, NULL_POINTER, OTHER_EXCEPTION)
     */
    public synchronized int checkExitValue() {
	try {
	    return Math.abs(getProcess().exitValue());
	}
	catch (IllegalThreadStateException  e) {
	    return ILLEGAL_THREAD_STATE;
	}
	catch (NullPointerException  e) {
	    return NULL_POINTER;
	}
	catch (Exception e) {
	    return OTHER_EXCEPTION;
	}
    }
	
    /**
     * This function returns the exit value of the running process. If the process is not finished, this method 
     * will block until it finishes, or the TestObject timeout value is reached.
     * If the test times-out, it is explicitly killed, as if a ACTION_TESTOBJECT_KILL message was sent by the harness.
     * @return the exit value of the process.
     */
    public synchronized int getExitValue() {
	if (exitValue>=0) {
	    ConsoleServer.debugMsg("Process already finished ("+exitValue+") - returning previous exit code",9);
	    return exitValue;
	}
	else {
	    while ((!wasInterrupted())&&
		   ((exitValue = checkExitValue())<0)&&
		   (!testTimedOut())&&
		   (exitValue>=OTHER_EXCEPTION)) {
		ConsoleServer.debugMsg("Checking process exit value ("+exitValueString(exitValue)+") - will timeout in "+getTimeLeft()+" seconds",9);
		Utils.safeSleep(1);
	    }
	    // check if it timed out
	    if (testTimedOut())
		exitValue = TIMEDOUT_STATE;
	    return exitValue;
	}
    }
	
    /**
     * If the val is an internal QAT code, return a string version of the code,
     * else just return the integer value as a String.
     */
    private String exitValueString(int val) {
	if (val < 0) {
	    switch (val) {
	    case NOT_YET_SET         :return "NOT_YET_SET";
	    case ILLEGAL_THREAD_STATE:return "NOT_FINISHED";
	    case TIMEDOUT_STATE      :return "TIMEDOUT_STATE";
	    case NULL_POINTER        :return "NULL_POINTER";
	    case OTHER_EXCEPTION     :return "OTHER_EXCEPTION";
	    }
	}
	return Integer.toString(val);
    }
	
    /**
     * Returns true if we need to stop this process.
     */
    private boolean wasInterrupted() {
	return interrupted;
    }
	
    private void setInterrupted(boolean i) {
	interrupted = i;
    }
	
    private long getTimeLeft() {
	return ((startTime.getTime()+1000*test.getTimeout())-(new Date()).getTime())/1000;
    }
	
    /**
     * Will return false until the test has timed out, and then it will return true.
     * A test is considered as timed out if it runs longer than the value of TestObject.geTimeout() which
     * is measured in seconds.
     */
    private boolean testTimedOut() {
	Date now = new Date();
	if ((test.getTimeout()!=0)&&(now.after(new Date(startTime.getTime()+1000*test.getTimeout())))) {
	    return true;
	}
	else {
	    return false;
	}
    }
    /** 
     * This will kill the TestObject execution and free all the resource associated with this
     * TestObject execution.
     */
    public void interrupt() {
	setInterrupted(true);
		
	if (exitValue==NOT_YET_SET)
	    exitValue = checkExitValue();
		
	destroy();
		
	if (stdin != null) {
	    stdin.interrupt();
	    stdin = null;
	}
	if (stderr != null) {
	    stderr.interrupt();
	    stderr = null;
	}
    }
	
    /**
     * Outputs a debug message if the debug level corresponds to the level
     * of debug the agent is set to.
     */
    private void debugMsg(String msg, PrintWriter out, int level) {
	ConsoleServer.debugMsg(msg,level);
	try {
	    out.println(msg);
	}
	catch (Exception e) {
	    ConsoleServer.debugMsg("Error recording trace file :"+e,2);
	}
    }
	
    public boolean isDaemon() {
	return isDaemon;
    }
	
    public void setDaemon(boolean isDaemon) {
	this.isDaemon = isDaemon;
    }
	
    public void destroy() {
	Process proc = getProcess();
	if (proc != null)
	    proc.destroy();
    }

    }

File: src/chorus/ExecProcess.cpp
/*
 * @author webhiker
 * @version 1.0, 26 April 2000
 *
 */

#include "ExecProcess.hpp"
  

ExecProcess::ExecProcess(TestObject *t, 
						 int level, 
						 int mode) {
  test = t;
  debugLevel = level;
  showOutput = mode;
  setDaemon(FALSE);  
}

ExecProcess::~ExecProcess()
{	
  if (test!=NULL) {
	delete test;	
	test = NULL;
  }
}

/** 
	 * This method returns the TestObject associated with this Object.
	 * @return the TestObject associated with this object.
	 */
TestObject *ExecProcess::getTestObject() {
  return test;
}

/** 
 * This method should not be called directly, but rather via ObjectName.start() to
 * ensure the thread is started correctly.
 */
void ExecProcess::run() {
  
  string **argsString;	
  string **envString;

  argsString = test->getExecuteCommand();
  envString = test->getExecuteEnv();
	
  exitValue = NOT_YET_SET;
  int process;
  int pidChild;
	
  // create our env file as well
  ConsoleServer::debugMsg(1,"Opening environment file %s\n\n",test->getEnvFileName().c_str());
  try {
	FILE *envFile = Utils::createF(test->getEnvFileName());		
		
	ConsoleServer::debugMsgF(1,envFile,"Agent details :\n");
	ConsoleServer::debugMsgF(1,envFile,"---------------\n");
	ConsoleServer::debugMsgF(1,envFile,"Host name\n");
	ConsoleServer::debugMsgF(1,envFile,"\n",1);
		
	ConsoleServer::debugMsgF(1,envFile,"Command :\n");
	ConsoleServer::debugMsgF(1,envFile,"---------------------\n");	
	for (int i = 0; i < (test->getCommandCount()); i++) {	  
	  ConsoleServer::debugMsgF(1,envFile,"Argument %d is %s\n",i,argsString[i]->c_str());
	}
	ConsoleServer::debugMsgF(1,envFile,"\n");
		
	ConsoleServer::debugMsgF(1,envFile,"Process environment :%d\n",test->getExecuteEnvCount());
	ConsoleServer::debugMsgF(1,envFile,"---------------------\n");
	for (int i = 0; i < test->getExecuteEnvCount(); i++) {	  
	  ConsoleServer::debugMsgF(1,envFile,"%s\n",envString[i]->c_str());	  
	}
		
	ConsoleServer::debugMsgF(1,envFile,"\n");
		
	pidChild = fork(); 

	if (pidChild == -1) {

	  ConsoleServer::debugMsg(1,"Failed to fork :%s errno=%d\n", strerror(errno),errno);
	}
	else{
	  if (pidChild==0) {
		// this is the child process		
		// so redirect it's stdout and stderr to file			
		redirectOutput(envFile);	
		
		// create the prog_args and prog_env arrays
		char **prog_args = new char*[test->getCommandCount()+1];
		for (int i = 0; i < test->getCommandCount(); i++) {
		  prog_args[i] = (char *)argsString[i]->c_str();		  
		}
		prog_args[test->getCommandCount()]=NULL;
		
		char **prog_env = new char*[test->getExecuteEnvCount()+1];
		for (int i = 0; i < test->getExecuteEnvCount(); i++) {
		  prog_env[i] = (char *)envString[i]->c_str();		  
		}
		prog_env[test->getExecuteEnvCount()]=NULL;
		
		process = execve(argsString[0]->c_str(),		
				 (char*const*)prog_args,
				 (char*const*)prog_env); 
		if ( process < 0 ) {		  
		  ConsoleServer::debugMsg(1,"Error starting process :%s\n",strerror(errno));
		}
		
		// clean up prog_args
		delete prog_args;
		// clean up env
		delete prog_env;

		exit(process);			
	  }
	  else {			
		// this is the parent process
		pid = pidChild;	
		startTime = time(0);
				
		char *timebuff = Utils::allocateChars(STR_LENGTH);	
		timebuff = ctime_r(&startTime,timebuff);			
		ConsoleServer::debugMsgF(1,envFile,"Process %d started at %s\n",pidChild,timebuff);
		delete [] timebuff;			
		timebuff = NULL;
				
		fclose(envFile);
	  }
	}
  }
  catch (char * message) {
	ConsoleServer::debugMsg(1,"Error trying to run the process :%s\n",message);
  }
	
}

/**
 * Indicates the process to start will be a daemon, and so don't bother 
 * keeping it's traces.
 */
void ExecProcess::setDaemon(int isd) 
{
  isADaemon = isd;  
}

int ExecProcess::isDaemon() 
{
  return isADaemon;
}


void ExecProcess::redirectOutput(FILE *envFile) 
{		  		  		
  ConsoleServer::debugMsgF(3,envFile,"Opening stderr file:%s\n",test->getStdErrFileName().c_str());
  dup2( Utils::createH(test->getStdErrFileName()),STDERR_FILENO);
		
  ConsoleServer::debugMsgF(3,envFile,"Opening stdout file:%s\n",test->getStdOutFileName().c_str());
  dup2(Utils::createH(test->getStdOutFileName()),STDOUT_FILENO);
}

/**
	 * This function returns the exit value of the running process. If the process is not finished, this method 
	 * will block until it finishes, or the TestObject timeout value is reached.
	 * If the test times-out, it is explicitly killed, as if a ACTION_TESTOBJECT_KILL message was sent by the harness.
	 * @return the exit value of the process.
	 */
int ExecProcess::getExitValue() {
  if (exitValue!=NOT_YET_SET) {
	ConsoleServer::debugMsg(9,"Process already finished (%d) - returning previous exit code\n",exitValue);
  }
  else {
	while ((!testTimedOut())&
		   (stillRunning(pid))) {
	  switch (exitValue) {
	  case NOT_YET_SET:
		ConsoleServer::debugMsg(9,"Checking process exit value (%s) - will timeout in %d seconds\n",
								"Still running",
								getTimeLeft());
		break;
	  case TIMEDOUT_STATE :
		ConsoleServer::debugMsg(9,"Checking process exit value (%s) - will timeout in %d seconds\n",
								"Timed out",
								getTimeLeft());
		break;
	  }		
	  sleep(1);
	}
	// exit value set by either testTimedOut() or stillRunning()
  }
  return exitValue;
}

/**
 * Returns the exit value of the process if it has finished, else
 * it returns a negative value. Will not block.
 */
int ExecProcess::checkExitValue()   
{
  // call this method to set up the correct value
  stillRunning(pid);
  // return the result
  return exitValue;  
}

long ExecProcess::getTimeLeft() {
  time_t now = time(0);
  time_t left = (test->getTimeout()-(now-startTime));
  if (left>0)
	return left;
  else
	return 0;	
}
	
/**
	 * Will return false until the test has timed out, and then it will return true.
	 * A test is considered as timed out if it runs longer than the value of TestObject.geTimeout() which
	 * is measured in seconds.
	 */
int ExecProcess::testTimedOut() {
  if (test==NULL) {
	return TRUE;
  }
  
  if ((test->getTimeout()!=0)
	  &(getTimeLeft()==0)) {
	interrupt();
	return TRUE;
  }
  else {
	return FALSE;
  }
}

int ExecProcess::stillRunning(pid_t child_pid) 
{
  if (exitValue==NOT_YET_SET){
	int stat_loc;	
	int result;

	result = waitpid(child_pid, &stat_loc, WNOHANG);

	if (result==0){	  
	  return TRUE;				
	}
	else {
	  if (result==child_pid) {
		if (WIFSIGNALED(stat_loc))
		  exitValue = WTERMSIG(stat_loc);
		else
		  exitValue = WEXITSTATUS(stat_loc);
		return FALSE;		
	  }
	  else {
		ConsoleServer::debugMsg(1,"Error retrieving exit value :%s\n",strerror(errno));
		exitValue = NOT_YET_SET;
		return FALSE;				
	  }
	}		
  }
  else {
	return FALSE;
  }
}

/** 
 * This will kill the TestObject execution and delete all the resource associated with this
 * TestObject execution.
 */
void ExecProcess::interrupt() {
  if (stillRunning(pid)) {
	ConsoleServer::debugMsg(1,"Interrupting process %d\n",pid);

	ConsoleServer::debugMsg(1,"Regular ACTOR Kill requested\n");
	if (kill(pid,SIGTERM)!=0) {	  
	  ConsoleServer::debugMsg(1,"Process %d did not respond to kill SIGTERM %d request - trying harder\n",pid,SIGTERM);
	  if (kill(pid,SIGKILL)!=0) {	  
		ConsoleServer::debugMsg(1,"Warning: Process %d did not respond to kill SIGKILL %d - may still be running\n",pid,SIGKILL);
	  }
	}
	exitValue = TIMEDOUT_STATE;	
  }
  else {
	ConsoleServer::debugMsg(1,"Process %d did not appear to be running (already finished?)\n",pid);
  }
}

pid_t ExecProcess::getPid() {
  return pid;  
}
